### PR TITLE
feat: Add instrumentation config validation

### PR DESCRIPTION
**Pull request by robertlaurin**
During this weeks SIG I proposed adding a validation step of sorts to create a more structured interface for our instrumentation configuration.

This idea sparked from this comment chain https://github.com/open-telemetry/opentelemetry-ruby/pull/594#discussion_r568753846

### CODE REVIEW

**Comment by robertlaurin** [comment]
Hey @mwear & @ericmustin wanted to get some feedback on this approach before I keep going with it.  I used the graphql instrumentation as an example of what the configuration options would look like.

File: api/lib/opentelemetry/instrumentation/base.rb [hunk 1]
```diff
@@ -180,12 +186,31 @@ def install(config = {})
...
+      def config_options(config_hash)
+        config_hash ||= {}
+
+        @options.each_with_object({}) do |option, h|
+          option_name = option[:name]
+          config_hash_value = config_hash[option_name]
+
+          value = if config_hash_value.nil? && option[:allow_nil]
+            nil
+          elsif option[:validate].call(config_hash_value)
```
Comment (ericmustin) [inline api/lib/opentelemetry/instrumentation/base.rb:204]: In the same way we're providing a validation of config options here, we  shouuld probably validate that this is callable, or a proc, or something to that effect? Wdyt?

File: api/lib/opentelemetry/instrumentation/base.rb [hunk 2]
```diff
@@ -180,12 +186,31 @@ def install(config = {})
...
+            config_hash_value
```
Comment (ericmustin) [inline api/lib/opentelemetry/instrumentation/base.rb:205]: So is the idea that the validation proc itself would log a warning? or, should we attempt to log something uniform here?

**Review by ericmustin — COMMENTED** [review]
I generally think having a `validate` option is reasonable, the only thing i'd say is that we probably want to have some expected failure behavior here, and try to standardize that, rather than letting the validation proc itself be in charge (imo validation proc should just return true/false).

Specifically, i think have a `warn` level log message with
- the integration name, 
- option name, 
- invalid value, 
- and fallback value

Would be appropriate and have enough information to let users fix their inputs without spamming their logs. that's just my two cents tho, my real concern here is that leaving it up to the specific integration's validation proc will create logs at different levels and different amounts of detail over time as integrations get added.


Additionally, from an implementation standpoint, I think we'd need to be defensive and make sure

1. the validate option itself is a proc
2. we catch/swallow any exceptions that get raised within the validation proc.

File: api/lib/opentelemetry/instrumentation/base.rb (hunk 1 above)
Comment (robertlaurin) [reply api/lib/opentelemetry/instrumentation/base.rb:204]: I think that's a reasonable thing to do, tests would only catch that error if there is test explicitly setting that config option.

File: api/lib/opentelemetry/instrumentation/base.rb (hunk 2 above)
Comment (robertlaurin) [reply api/lib/opentelemetry/instrumentation/base.rb:205]: >imo validation proc should just return true/false
I agree with what you mentioned above that the proc should only return true/false.  I think that we should only be logging an error/warning/info (to be discussed) in the event that the validation fails on user provided config options.  

When no option is provided and the default is used I think that's a quiet happy path.

Comment (robertlaurin) [reply api/lib/opentelemetry/instrumentation/base.rb:205]: > should we attempt to log something uniform here?

I believe we should be responsible for the logging and not the proc.

File: instrumentation/rack/lib/opentelemetry/instrumentation/rack/instrumentation.rb [hunk 3]
```diff
@@ -22,6 +22,14 @@ class Instrumentation < OpenTelemetry::Instrumentation::Base
           defined?(::Rack)
         end
 
+        option :allowed_request_headers,  default: [],    validate: ->(v) { v.is_a?(Array) }
+        option :allowed_response_headers, default: [],    validate: ->(v) { v.is_a?(Array) }
+        option :application,              default: nil,   validate: ->(v) { v.respond_to?(:call) }
+        option :record_frontend_span,     default: false, validate: ->(v) { v.is_a?(TrueClass) || v.is_a?(FalseClass) }
+        option :retain_middleware_names,  default: false, validate: ->(v) { v }
```
Comment (fbogsany) [inline instrumentation/rack/lib/opentelemetry/instrumentation/rack/instrumentation.rb:29]: What's the deciding factor for choosing
```ruby
default: false, validate: ->(v) { v.is_a?(TrueClass) || v.is_a?(FalseClass) }
```
vs
```ruby
default: false, validate: ->(v) { v }
```

File: api/lib/opentelemetry/instrumentation/base.rb [hunk 4]
```diff
@@ -225,6 +239,39 @@ def enabled?(config = nil)
 
       private
 
+      # The config_options method is responsible for validating that the user supplied
+      # config hash is valid.
+      # Unknown configuration keys are not included in the final config hash.
```
Comment (fbogsany) [inline api/lib/opentelemetry/instrumentation/base.rb:244]: Should we log unknown configuration keys?

File: api/lib/opentelemetry/instrumentation/base.rb [hunk 5]
```diff
@@ -129,14 +129,27 @@ def compatible(&blk)
           @compatible_blk = blk
         end
 
+        # The option method is used to define default configuration options
+        # for the instrumentation library. It requires a name, default value,
+        # and a validation callable to be provided.
+        # @param [String] name The name of the configuration option
+        # @param default The default value to be used, or to used if validation fails
+        # @param [Callable] validate A callable that validates the option value and returns a boolean
```
Comment (fbogsany) [inline api/lib/opentelemetry/instrumentation/base.rb:137]: We may want to relax this at some point and make the `validate` Callable optional, but I think it's fine to keep this restrictive for now.

File: instrumentation/sidekiq/lib/opentelemetry/instrumentation/sidekiq/instrumentation.rb [hunk 6]
```diff
@@ -27,6 +27,12 @@ class Instrumentation < OpenTelemetry::Instrumentation::Base
           gem_version >= MINIMUM_VERSION
         end
 
+        option :enable_job_class_span_names, default: false, validate: ->(v) { v.is_a?(TrueClass) || v.is_a?(FalseClass) }
```
Comment (fbogsany) [inline instrumentation/sidekiq/lib/opentelemetry/instrumentation/sidekiq/instrumentation.rb:30]: Is this sufficient or do we need to explicitly test the class?
```suggestion
        option :enable_job_class_span_names, default: false, validate: ->(v) { v == true || v == false }
```

File: api/lib/opentelemetry/instrumentation/base.rb (hunk 4 above)
Comment (robertlaurin) [reply api/lib/opentelemetry/instrumentation/base.rb:244]: Yeah I think so, it seems like a helpful addition to anyone getting setup with an instrumentation library.

File: instrumentation/rack/lib/opentelemetry/instrumentation/rack/instrumentation.rb (hunk 3 above)
Comment (robertlaurin) [reply instrumentation/rack/lib/opentelemetry/instrumentation/rack/instrumentation.rb:29]: Ah, this was just some leftover.  Will make it consistent.

File: instrumentation/sidekiq/lib/opentelemetry/instrumentation/sidekiq/instrumentation.rb (hunk 6 above)
Comment (robertlaurin) [reply instrumentation/sidekiq/lib/opentelemetry/instrumentation/sidekiq/instrumentation.rb:30]: I don't think we need to test the class.  I'll update it.

**Comment by robertlaurin** [comment]
Made a few changes based on feedback, the only thing left from my perspective would be potentially adding some common validation helpers for booleans, integers, arrays, etc... 

I'm not sure where they would be best suited to exist, my first thought would be having them be in the `common` gem.

**Review by fbogsany — APPROVED** [review]
This LGTM. I'm inclined to 🚢 as is and refine it with helpers later (whether that's relaxing the `validate:` argument to accept symbols or packaging helpers somewhere). If we're going to accept symbols, then the helpers should be in the API gem. Otherwise they should be added to `opentelemetry-common`.

**Comment by fbogsany** [comment]
Hmm, those Rubocop failures are annoying. I don't think we want to encourage ignoring them everywhere, so maybe we should add those helpers now 😄. I like the symbol approach:
```ruby
option :enable_job_class_span_names, default: false, validate: :boolean
```

**Review by ericmustin — APPROVED** [review]
🚀

File: api/lib/opentelemetry/instrumentation/base.rb [hunk 7]
```diff
@@ -59,9 +59,17 @@ module Instrumentation
     # convention for environment variable name is the library name, upcased with
     # '::' replaced by underscores, OPENTELEMETRY shortened to OTEL_{LANG}, and '_ENABLED' appended.
     # For example: OTEL_RUBY_INSTRUMENTATION_SINATRA_ENABLED = false.
-    class Base
+    class Base # rubocop:disable Metrics/ClassLength
       class << self
         NAME_REGEX = /^(?:(?<namespace>[a-zA-Z0-9_:]+):{2})?(?<classname>[a-zA-Z0-9_]+)$/.freeze
+        VALIDATORS = {
+          array: ->(v) { v.is_a?(Array) },
+          boolean: ->(v) { v.is_a?(TrueClass) || v.is_a?(FalseClass) },
```
Comment (fbogsany) [inline api/lib/opentelemetry/instrumentation/base.rb:67]: _sigh_ I almost think it's worth adding a `# rubocop:disable` to this line, but not quite 🤷

File: api/lib/opentelemetry/instrumentation/base.rb (hunk 5 above)
Comment (fbogsany) [reply api/lib/opentelemetry/instrumentation/base.rb:145]: This comment should be expanded to include the built-in validators.

File: api/lib/opentelemetry/instrumentation/base.rb [hunk 8]
```diff
@@ -59,9 +59,17 @@ module Instrumentation
...
+          callable: ->(v) { v.respond_to?(:call) },
+          integer: ->(v) { v.is_a?(Integer) },
+          string: ->(v) { v.is_a?(String) }
+        }.freeze
+
         private_constant :NAME_REGEX
```
Comment (fbogsany) [inline api/lib/opentelemetry/instrumentation/base.rb:73]:
```suggestion
        private_constant :NAME_REGEX, :VALIDATORS
```

**Comment by robertlaurin** [comment]
@fbogsany got those two comments fixed up!

**Review by fbogsany — APPROVED** [review]

**Review by mwear — APPROVED** [review]
This is a big improvement! Thanks @robertlaurin!

**Pull request merged**